scripts/py_matter_idl/matter_idl/generators/java/ChipClusters_java.jinja

```diff
@@ -126,22 +180,64 @@ public class ChipClusters {
      * empty optional, the default timeout will be used.
      */
     public void setCommandTimeout(Optional<Long> timeoutMillis) {
-      setCommandTimeout(chipClusterPtr, timeoutMillis);
+      this.timeoutMillis = timeoutMillis;
     }
 
-    private native void setCommandTimeout(long clusterPtr, Optional<Long> timeoutMillis);
-
     /** Returns the current timeout (in milliseconds) for commands sent through this cluster. */
     public Optional<Long> getCommandTimeout() {
-      Optional<Long> timeout = getCommandTimeout(chipClusterPtr);
-      return timeout == null ? Optional.empty() : timeout;
+      return timeoutMillis == null ? Optional.empty() : timeoutMillis;
     }
 
-    private native Optional<Long> getCommandTimeout(long clusterPtr);
-
+    @Deprecated
     public abstract long initWithDevice(long devicePtr, int endpointId);
 
-    public native void deleteCluster(long chipClusterPtr);
+    protected void readAttribute(
+        ReportCallbackImpl callback,
+        long attributeId,
+        boolean isFabricFiltered) {
+      ReportCallbackJni jniCallback = new ReportCallbackJni(null, callback, null);
+      ChipAttributePath path = ChipAttributePath.newInstance(endpointId, clusterId, attributeId);
+      ChipDeviceController.read(0, jniCallback.getCallbackHandle(), devicePtr, Arrays.asList(path), null, isFabricFiltered, timeoutMillis.orElse(0L).intValue(), null);
+    }
+
+    protected void writeAttribute(
+        WriteAttributesCallbackImpl callback,
+        long attributeId,
+        BaseTLVType value,
+        int timedRequestTimeoutMs) {
+      WriteAttributesCallbackJni jniCallback = new WriteAttributesCallbackJni(callback);
+      byte[] tlv = encodeToTlv(value);
```

Comment:
If value is a primitive type, this will encode the primitive type with AnonymousTag. Based on your implementation, it looks:

1. For invoke command
It expects the top-level TLV element to be a structure with an AnonymousTag.

2. To write a attribute with primitive type
It expects the top-level TLV element to be a primitive with an AnonymousTag.

3. To write a attribute with structure type
It expects the top-level TLV element to be a structure with an AnonymousTag.

4. For the TLV payload in invoke/read/subscribe response
It expects the top-level TLV element to be a structure with an AnonymousTag.

Is my understanding correct? I thought all the top level is tag with anonymous struct, no?

Reply:
yes